### PR TITLE
Add an ActiveMQ Artemis extension

### PR DESCRIPTION
**Pull request by middagj**
I have added an extension to support ActiveMQ Artemis via Core or JMS. Using AMQP via reactive messaging does not suit our needs as we want to control the acknowledgements and commits.

I have kept the CDI part small and only produce a ServerLocater (Core) and ActiveMQJMSConnectionFactory (JMS). These components should not be created again even if the connection dies for some reason. I don't know if building a framework around this (using an annotation for the listener method, create a proper producing template with serialization and transactions) is needed. It would take considerably more time, but would definitely add developer joy.

I also have chosen to support a single url property and not all the options separately. In Artemis you can specify the options there (like JDBC).

The changes for the Netty native libraries I have put in the NettyProcessor. They postpone initialization of those classes to runtime. I have disable epoll and kqueue in SubstrateVM for Artemis because of an UnsatisfiedLinkError. Others also seem to fail, e.g. [Ratpack](https://e.printstacktrace.blog/ratpack-graalvm-how-to-start/), [Vert.x](https://github.com/oracle/graal/issues/442).

### CODE REVIEW

**Comment by middagj** [comment]
```##[Error 1]
The job running on agent Hosted Agent has exceeded the maximum execution time of 150.
```
Is this normal? A quick inspection in the log shows that Maven downloads at very low transfer rates.

**Comment by gsmet** [comment]
@middagj don't worry about it, we need to set the timeout higher.

**Comment by geoand** [comment]
I already uped the timeout as part of #2910

**Comment by geoand** [comment]
@middagj could you please rebase onto the latest master to fix the conflicts?

**Comment by gastaldi** [comment]
I wonder if this PR solves https://github.com/quarkusio/quarkus/issues/1843?

**Comment by middagj** [comment]
I have rebased on master. Yes this would solve #1843, but the reporter should still manually enable JNI.

**Comment by geoand** [comment]
@gsmet perhaps we need to add a BuildItem that enables JNI?

**Comment by cescoffier** [comment]
@middagj Just to clarify, Reactive Messaging gives you total control on the acknowledgment using `Message.ack()`:

```
CompletionStage<Void> consume(Message<String> message) {
   // do something with the message
  // acknoweldge
  return message.ack();
}
```

In addition, did you look at the impact on the native executable size of the runtime initialization items?

**Comment by middagj** [comment]
Thanks, didn't know that was possible. However, I don't see transactions being possible in the reactive messaging framework, which we would like to use.

I did not look at the impact on the native executable size. Just did a comparison by disabling the runtime initializations and excluding the epoll and kqueue libraries explicitly (needed the `--allow-incomplete-classpath` flag) vs the proposed solution . On a small working example the difference is around 340kB on 58MB. But this will disable epoll and kqueue also when not running native but in the VM and gives NoClassDefFoundError exceptions when running.

**Comment by emmanuelbernard** [comment]
@cescoffier do you mind if I make you captain for that PR review. You're probably the one with the best view on how we want to proceed here.

**Comment by middagj** [comment]
I don't know if the label reactive messaging really fits this PR. My goal of this PR is to add support for the Artemis Core/JMS client in Quarkus. One could already use Artemis with the reactive messaging framework via AMQP, as already demonstrated in the AMQP guide.

File: integration-tests/artemis-core/src/main/java/io/quarkus/it/artemis/ArtemisProducerManager.java
Comment (emmanuelbernard) [inline integration-tests/artemis-core/src/main/java/io/quarkus/it/artemis/ArtemisProducerManager.java]: I'm not quite sure whyw e don't enlist a CDI bean being the session factory provider in the extension

File: extensions/infinispan-client/deployment/src/main/java/io/quarkus/infinispan/client/deployment/InfinispanClientProcessor.java
Comment (emmanuelbernard) [inline extensions/infinispan-client/deployment/src/main/java/io/quarkus/infinispan/client/deployment/InfinispanClientProcessor.java]: Found issue with Infinispan?

File: extensions/artemis-core/runtime/src/main/java/io/quarkus/artemis/core/runtime/ArtemisBuildConfig.java
Comment (emmanuelbernard) [inline extensions/artemis-core/runtime/src/main/java/io/quarkus/artemis/core/runtime/ArtemisBuildConfig.java]: Not a big fan of these runtime discovery in Quarkus. Any reasons why the JMS extension could not send a build step that can be listened to by artemis core to decide at build time instead of runtime?

File: integration-tests/artemis-core/src/main/java/io/quarkus/it/artemis/ArtemisProducerManager.java
Comment (middagj) [reply integration-tests/artemis-core/src/main/java/io/quarkus/it/artemis/ArtemisProducerManager.java]: Because a ClientSessionFactory should be regarded as connection. See [Artemis documentation](https://activemq.apache.org/components/artemis/documentation/latest/core.html#clientsessionfactory). If your connection breaks, the ClientSessionFactory should be created again.

File: extensions/infinispan-client/deployment/src/main/java/io/quarkus/infinispan/client/deployment/InfinispanClientProcessor.java
Comment (middagj) [reply extensions/infinispan-client/deployment/src/main/java/io/quarkus/infinispan/client/deployment/InfinispanClientProcessor.java]: Yes, the reason is the added RuntimeInitializedClass of `io.netty.channel.epoll.Epoll` in the NettyProcessor [1]. Although the `TransportHelper` is being substituted in [SubstituteTransportHelper](https://github.com/quarkusio/quarkus/blob/master/extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/graal/SubstituteTransportHelper.java) where there isn't a dependency on Epoll, the original [TransportHelper](https://github.com/infinispan/infinispan/blob/master/client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/transport/netty/TransportHelper.java) does have this dependency. When building a native image I got an exception that TransportHelper was dependent on Epoll which was runtime initialized. It was a separated commit, but I squashed in the rebase. Although when running it again today without this addition, it didn't seem necessary anymore.

[1] I thought it would be best to put it in the NettyProcessor, then to put it in the ArtemisProcessor. I don't know if there is an convention, some libraries tend to put their Netty code in their own library. For example the InfinispanClientProcessor has `SystemPropertyBuildItem("io.netty.noUnsafe", "true")`, but so does the NettyProcessor.

File: extensions/artemis-core/runtime/src/main/java/io/quarkus/artemis/core/runtime/ArtemisBuildConfig.java
Comment (middagj) [reply extensions/artemis-core/runtime/src/main/java/io/quarkus/artemis/core/runtime/ArtemisBuildConfig.java]: I am not a big fan myself either (although it is only being used build time). Have you a good example in the code how to do this? I will add something. And would you ditch the option? I was doubting, because I think if you include artemis-jms (which depends on artemis-core as the JMS client is a wrapper around their core client) you probably want to use the JMS client.

**Comment by emmanuelbernard** [comment]
Yes don't worry too much about the label name here.

File: integration-tests/artemis-core/src/main/java/io/quarkus/it/artemis/ArtemisProducerManager.java
Comment (emmanuelbernard) [reply integration-tests/artemis-core/src/main/java/io/quarkus/it/artemis/ArtemisProducerManager.java]: @middagj OK but that feels like a terrible developer experience.
The extension should make an effort to retry this connection itself. You can imagine very few users would think of doing it otherwise. Any advice on how to do it with CDI @mkouba

File: extensions/infinispan-client/deployment/src/main/java/io/quarkus/infinispan/client/deployment/InfinispanClientProcessor.java
Comment (emmanuelbernard) [reply extensions/infinispan-client/deployment/src/main/java/io/quarkus/infinispan/client/deployment/InfinispanClientProcessor.java]: I forgot if there is a best practice @stuartwdouglas 
Ideally if that's "statically" shared better put it in the based extension.

File: extensions/artemis-core/runtime/src/main/java/io/quarkus/artemis/core/runtime/ArtemisBuildConfig.java
Comment (emmanuelbernard) [reply extensions/artemis-core/runtime/src/main/java/io/quarkus/artemis/core/runtime/ArtemisBuildConfig.java]: Look at this. Agroal is the connection pool that publishes a datasource driver

https://github.com/quarkusio/quarkus/blob/master/extensions/agroal/deployment/src/main/java/io/quarkus/agroal/deployment/AgroalProcessor.java#L71
https://github.com/quarkusio/quarkus/blob/master/extensions/agroal/deployment/src/main/java/io/quarkus/agroal/deployment/AgroalProcessor.java#L104

And some listen to these build items.
https://github.com/quarkusio/quarkus/blob/master/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java#L120

File: integration-tests/artemis-core/src/main/java/io/quarkus/it/artemis/ArtemisProducerManager.java
Comment (mkouba) [reply integration-tests/artemis-core/src/main/java/io/quarkus/it/artemis/ArtemisProducerManager.java]: Well, if I understand it correctly we could create bean that implements `ClientSessionFactory` and delegates to a factory created by `serverLocator.createSessionFactory()`. Then in every call we could verify the underlying factory and if something is wrong, recreate the factory again? @middagj @emmanuelbernard But I don't know anything about Artemis. So this is just a guess.

Comment (middagj) [reply integration-tests/artemis-core/src/main/java/io/quarkus/it/artemis/ArtemisProducerManager.java]: I think the developer joy can be greatly improved. I would like to not have to worry about a connection at all and just want to define a message listener and have a producer bean which can be used to send new messages in the same transaction if configured. I don't know when all that would realistic to add on, and having the ability to use Artemis client in Quarkus already helps us and probably others already.

Sounds like a good solution. There are different session factories which can be created, but that can be put in a config.

**Comment by emmanuelbernard** [comment]
That's my understanding too Martin

File: extensions/artemis-jms/runtime/src/main/java/io/quarkus/artemis/jms/runtime/ArtemisJmsProducer.java
Comment (cescoffier) [inline extensions/artemis-jms/runtime/src/main/java/io/quarkus/artemis/jms/runtime/ArtemisJmsProducer.java]: Shouldn't it be JMSConnectionFactory? The fact that it's ActiveMQ is an implementation detail for the user; or maybe I'm missing something.

**Review by cescoffier — CHANGES_REQUESTED** [review]
I've made one small suggestion.

The rest looks good. I would like to have the opinion of @dmlloyd and @stuartwdouglas on the `Class.forName` and the runtime initialization.

Side notes:
1. As you are exposing the ConnectionFactory, I can now implement the JMS support in reactive messaging which will use the produced ConnectionFactory
2. About transaction, it's coming, but far from being perfect as XA is a blocking protocol. Right now the solution is to off-load to a thread pool. But nevertheless, the JMS connector need a ConnectionFactory and I can use yours.

File: extensions/artemis-jms/runtime/src/main/java/io/quarkus/artemis/jms/runtime/ArtemisJmsProducer.java
Comment (middagj) [reply extensions/artemis-jms/runtime/src/main/java/io/quarkus/artemis/jms/runtime/ArtemisJmsProducer.java]: If you like. I always like to return the specific type so that one could use that as well. You can still inject JMSConnection and it will use this one, at least that is how I used to work in Spring.

**Comment by middagj** [comment]
If I have time today I will see if I can give an alternative for the `Class.forName` as suggested by @emmanuelbernard.

Perhaps it is another topic, but XA transactions aren't necessary to use JMS transactions.

**Comment by dmlloyd** [comment]
Every single messaging and database API having its own transaction API is not a feature ;)

/cc @mmusgrov et al

**Comment by emmanuelbernard** [comment]
On Tue, Jun 25, 2019 at 4:15 PM David M. Lloyd <notifications@github.com>
wrote:

> Every single messaging and database API having its own transaction API is
> not a feature ;)
>
> /cc @mmusgrov <https://github.com/mmusgrov> et al
>
Why do you say that? Or rather to which part of that PR or discussion are
you referring to?

**Comment by dmlloyd** [comment]
> On Tue, Jun 25, 2019 at 4:15 PM David M. Lloyd ***@***.***> wrote: Every single messaging and database API having its own transaction API is not a feature ;) /cc @mmusgrov <https://github.com/mmusgrov> et al
> Why do you say that? Or rather to which part of that PR or discussion are you referring to?

I've just shamelessly co-opted the discussion to support a point in a completely separate conversation ;)

File: extensions/artemis-core/runtime/src/main/java/io/quarkus/artemis/core/runtime/ArtemisBuildConfig.java
Comment (middagj) [reply extensions/artemis-core/runtime/src/main/java/io/quarkus/artemis/core/runtime/ArtemisBuildConfig.java]: Thanks, I have added a commit that uses a BuildItem instead of `Class.forName`. I got rid of the BuildConfig, as I don't see any scenario where you going to include `artemis-jms` but still want to use the Core API.

File: extensions/artemis-jms/runtime/src/main/java/io/quarkus/artemis/jms/runtime/ArtemisJmsProducer.java
Comment (middagj) [reply extensions/artemis-jms/runtime/src/main/java/io/quarkus/artemis/jms/runtime/ArtemisJmsProducer.java]: You are right, if you want to use JMS, no need to use depend on the specific implementation.

File: extensions/infinispan-client/deployment/src/main/java/io/quarkus/infinispan/client/deployment/InfinispanClientProcessor.java
Comment (middagj) [reply extensions/infinispan-client/deployment/src/main/java/io/quarkus/infinispan/client/deployment/InfinispanClientProcessor.java]: I tested again without this. It works on macos (before rebase it didn't), but still not Linux.
```
com.oracle.svm.core.util.UserError$UserException: Class that is marked for delaying initialization to run time got initialized during image building: io.netty.channel.epoll.Epoll
```
or 
```
com.oracle.svm.core.util.UserError$UserException: Class that is marked for delaying initialization to run time got initialized during image building: io.netty.channel.unix.Errors
```

**Comment by dmlloyd** [comment]
This change now has conflicts.  Could you please rebase?

**Comment by middagj** [comment]
I have fixed the conflicts and also fixed the requested changes by @cescoffier, any things that should be added for this pull request to be merged?

File: integration-tests/artemis-core/pom.xml
Comment (gsmet) [inline integration-tests/artemis-core/pom.xml]: You shouldn't need that anymore. If you need JNI enabled, your extension should produce a `JniBuildItem`.

**Review by gsmet — CHANGES_REQUESTED** [review]
Added a comment about the availability of `JniBuildItem`.

File: integration-tests/artemis-core/pom.xml
Comment (middagj) [reply integration-tests/artemis-core/pom.xml]: Artemis depends on Netty which needs it, shouldn't it be better to create a separate pull request to fix that and rebase?

Comment (middagj) [reply integration-tests/artemis-core/pom.xml]: See #3267

Comment (gsmet) [reply integration-tests/artemis-core/pom.xml]: I merged it, you can remove the property from here too.

**Review by cescoffier — APPROVED** [review]

File: integration-tests/artemis-core/pom.xml
Comment (middagj) [reply integration-tests/artemis-core/pom.xml]: See #3286, added promised fix here. I probably need to rebase one after the other merges because of conflicts.

Comment (middagj) [reply integration-tests/artemis-core/pom.xml]: Fixed conflict

**Comment by middagj** [comment]
@gsmet I addressed your requested changes (and rebased again because of conflicts). Any other stuff that is blocking this pull request?

**Review by gsmet — APPROVED** [review]

File: integration-tests/artemis-core/src/main/java/io/quarkus/it/artemis/ArtemisProducerManager.java
Comment (middagj) [reply integration-tests/artemis-core/src/main/java/io/quarkus/it/artemis/ArtemisProducerManager.java]: I was trying this approach, but I reconsider if this is the right way. `ClientSessionFactory ` implements `AutoCloseable` and it resembles the normal JMS Connection class. If you look at both the [JMS](2930/files#diff-3e8b5119c05a5414d227f6917f4071d3) and [Core](2930/files#diff-c0cc000a32a46da7f908389e2861bbf9) ArtemisProducerManager, they look quite similar. Both protocols have the distinction between a Connection and a Session, only Core protocol names the Connection a SessionFactory.

I think it would be better to focus on JMS and create something similar like Spring's [JmsListener](https://docs.spring.io/spring/docs/5.0.x/spring-framework-reference/integration.html#jms-receiving) and [JmsTemplate](https://docs.spring.io/spring/docs/5.0.x/spring-framework-reference/integration.html#jms-sending) to improve developer joy. I prefer to do it in a separate pull request as this request already enables people to add the dependency in a native image.

**Comment by gsmet** [comment]
@cescoffier could you drive this puppy home?

We need a rebase but, apart from that, could you check if it's ready to go in?

**Comment by cescoffier** [comment]
@gsmet The rest was ok to be merged.

**Comment by middagj** [comment]
Did I forgot to add something that should result in the extension to be available in Maven central? I cannot find it after the release of 0.22.0.

**Comment by gsmet** [comment]
0.22.0 is 0.21.2 + the Spring extensions + bug fixes

This will be released as part of 0.23.0, which we hope to release soon.

**Comment by middagj** [comment]
Thanks!

**Comment by middagj** [comment]
Yes, it was on my todo list. I will add some today.

**Pull request merged**
